### DISCUSSION TITLE
Is there a way to do OAuth with Salesforce Input?

### DISCUSSION THREAD

**Discussion by zoomingrocket** [apache/hop#4653] [Q&A]
Hi,

New to Hop, our Salesforce Organization has disabled direct SOAP Login call which works based on username and password and only allows OAuth for all Authentication.

We have Salesforce Connected App & OAuth setup and can generate token from Postman. I can see that I can do a standalone REST client call to generate a usable bearer token, so can I subsequently interject this Bearer Token as SOAP Header "sessionID" to Salesforce Input or will I need to do my own Web Services Lookup or REST Client?
Salesforce Article for reference: https://developer.salesforce.com/blogs/developer-relations/2011/03/oauth-and-the-soap-api

Apologies in advance if this sounds like a stupid question :)

Thanks!

**Comment by bamaer** [comment]
First of all: it's not a stupid question, don't know why it would be.
We'll need to look into this in a bit more detail, but a lot of APIs accept tokens as passwords. If that works in the Salesforce transforms, the process of retrieving and updating a bearer token (afaik) won't work directly from those transforms, so that's something you'll have to take care of on your own for now. 
Feel free to create a feature request ticket for this, so we can keep track of it and hopefully implement it in a future release.

**Comment by zoomingrocket** [comment]
Captured feature request: Support OAuth Authentication for Salesforce Input/Output #4657

**Reply by bamaer** [reply]
thanks, nice and detailed!

**Comment by CatalyseRichard** [comment]
There is a way to do OAuth using the native javascript 'jsrsasign' libraries. We used it where we needed a JWT token in order to authenticate to a rest server.  
I don't have a generic example I can post but I'll try to find time to create one for you.  
In essence it goes as follows:

1. Obtain the jsrsasign library file and add it to a source-file folder
2. In a javascript transform use item0 tab to import the  js library.
- Note that you need variables to set both 'navigator' and 'window' values to "other" to ensure browser compatibility.
- Use 'LoadScriptFile(<jsrsa_file>) ; to import the library

3. The generic javascript for the main code tab (Script1) is attached here. It gets a token (JWT) and attaches it to a bearer which you would then use in the Rest call. Obviously, you'd need something to populate the input fields.

That's the most awkward part and works really well to get JWT tokens. You'd need other coding if you want to get other sorts of tokens - but the good thing is that the library will handle most OAuth methods.
[JWT_js_code.txt](https://github.com/user-attachments/files/18012534/JWT_js_code.txt)

Hope that helps a bit.

**Comment by hansva** [comment]
As of version 2.16, we have support for OAuth on the Salesforce transforms.

https://github.com/apache/hop/issues/4657
